nemo/utils/callbacks/s3_checkpoint_io.py

```diff
@@ -0,0 +1,272 @@
+import os
+import time
+from concurrent.futures import ProcessPoolExecutor
+from io import BytesIO
+from multiprocessing import get_start_method
+from pathlib import Path
+from tempfile import NamedTemporaryFile
+from typing import Any, Callable, Dict, Optional, Union
+
+import torch
+from lightning_fabric.plugins.io.checkpoint_io import CheckpointIO
+
+from nemo.utils import logging
+from nemo.utils.s3_utils import (
+    DEFAULT_CHUNK_SIZE_MB,
+    DEFAULT_MAX_READ_CONCURRENCY,
+    DEFAULT_MAX_WRITE_CONCURRENCY,
+    SHARED_MEM_DIR,
+    S3Utils,
+)
+
+
+class S3CheckpointIO(CheckpointIO):
+    """A custom S3CheckpointIO module that supports checkpoint reading/writing with s3 when filepath
```

Comment:
What kind of checkpoints is it compatible with?
Legacy MCore, distributed checkpoints MCore, FSDP?

Reply:
Legacy MCore format